lib/pos_fee_estimator.go

```diff
@@ -0,0 +1,506 @@
+}
+
+// estimateTxnFeeGivenTransactionRegister estimates the fee in nanos for the provided transaction
+// and transaction register using the congestionFactorBasisPoints, priorityPercentileBasisPoints,
+// and maxBlockSize params. It calls estimateFeeRateNanosPerKBGivenTransactionRegister to estimate
+// the fee rate and then computes the fee using computeFeeGivenTxnAndFeeRate.
+func (posFeeEstimator *PoSFeeEstimator) estimateTxnFeeGivenTransactionRegister(
+	txn *MsgDeSoTxn,
+	txnRegister *TransactionRegister,
+	congestionFactorBasisPoints uint64,
+	priorityPercentileBasisPoints uint64,
+	maxBlockSize uint64,
+) (uint64, error) {
+	feeRateNanosPerKB, err := posFeeEstimator.estimateFeeRateNanosPerKBGivenTransactionRegister(
+		txnRegister,
+		congestionFactorBasisPoints,
+		priorityPercentileBasisPoints,
+		maxBlockSize,
+	)
+	if err != nil {
+		return 0, errors.Wrap(err, "estimateTxnFeeGivenTransactionRegister: Problem computing fee rate")
+	}
+	txnFee, err := computeFeeGivenTxnAndFeeRate(txn, feeRateNanosPerKB)
+	if err != nil {
+		return 0, errors.Wrap(err, "estimateTxnFeeGivenTransactionRegister: Problem computing txn fee")
+	}
+	return txnFee, nil
+}
+
+// estimateFeeRateNanosPerKBGivenTransactionRegister estimates the fee rate in nanos per KB for the provided
+// transaction register and fee estimation parameters. The congestionFactorBasisPoints param is the congestion
+// factor in basis points (100 bps = 1%). The priorityPercentileBasisPoints param is the percentile of the
+// priority fee bucket to use for the fee rate estimation. The maxBlockSize param is the maximum block size
+// in bytes.
+// This estimates fee rates using the following approach:
+//  1. Compute the maximum size of numBlocks blocks as maxBlockSize * numBlocks, called maxSizeOfNumBlocks.
+//  2. Iterate over all transactions in the transaction register in fee time order until the total size of
+//     the transactions is greater than maxSizeOfNumBlocks and append those transactions to a slice.
+//  3. If there are no transactions in the slice after step 2, return the minimum network fee.
+//  4. Compute the priority fee bucket for the transactions in the slice using the priorityPercentileBasisPoints param
+//     by calling getPriorityFeeBucketFromTxns on the slice.
+//  5. If the resulting priority fee bucket from step 4 is less than the global minimum network fee, return the global
+//     minimum network fee.
+//  6. Compute the congestion threshold as congestionFactorBasisPoints * maxSizeOfNumBlocks / 10000.
+//  7. If the total size of the transactions in the slice is less than the congestion threshold, return one bucket lower
+//     than the priority fee bucket from step 4.
+//  8. If the total size of the transactions in the size is greater than the congestion threshold and less
+//     than or equal to maxSizeOfNumBlocks, return the priority fee bucket from step 4.
+//  9. Otherwise, return one bucket higher than the priority fee bucket from step 4.
```

Comment:
I'd link to the doc we have analyzing this stuff here.

Reply:
I would give some examples of these parameters here. I had to noodle pretty hard to work out examples and I'm the guy who wrote it! I think the examples below helped me:
```
When run for mempool:
numBlocks = 1 (this is set at construction time)
congestionFactorBasisPoints = 900 (90%)
priorityPercentileBasisPoints = 10000 (100%)

With these numbers, I think we'd have the following behavior (and notice I'm getting rid of the case where we keep the fee the same, see comment below):
- We will find the lowest fee paid by anyone within the first block's worth ot txns because priorityPercentileBasisPoints=100% will return the lowest-fee txn.
- If the mempool has >90% of 1 block's worth of txns, we will pay one fee bucket HIGHER than this lowest-fee txn.
- If the mempool has <90% of 1 block's worth of txns, we will pay one fee bucket LOWER than this lowest-fee txn.
```

More examples:
```
When run for mempool:
numBlocks = 5 (this is set at construction time)
congestionFactorBasisPoints = 10000 (100%)
priorityPercentileBasisPoints = 2000 (20%)

With these numbers, I think we'd have the following behavior (and notice I'm getting rid of the case where we keep the fee the same, see comment below):
- We find the fee of the 20th percentile txn, which is enough to be included in the next block
- If the mempool has more than 5 blocks worth of txns, we pay one fee bucket higher than this guy, otherwise we pay one fee bucket lower than him.

This one is a bit weird, and I think the simple single-block thing previously would be better but lmk what you think of it.
```

Now to play with some block fee estimation:
```
numBlocks = 5 (this is set at construction time)
congestionFactorBasisPoints = 9000 (90%)
priorityPercentileBasisPoints = 2000 (20%)

Here, we would find the 20th percentile txn, enough to include in the next block. And if the last 5 blocks are more than 90% full we pay 1 bucket higher than this guy. Otherwise, we pay one fee bucket lower.

This I think makes senes for the block fee estimator. Lmk if you agree.
```

Reply:
To summarize, I think the params we'd probably want for mempool and block estimator respectively are:

Mempool:
```
numBlocks = 1
congestionFactorBasisPoints = 9000 (90%)
priorityPercentileBasisPoints = 9000 (90%)

This means we will take one block's worth of txns from the mempool, and we'll pay 1 fee bucket higher than the 90th percentile guy if there's more than 90% of a block's worth of txns in it.
```

For the block estimator I think our params would be something like tihs:
```
numBlocks = 60
congestionFactorBasisPoints = 9000 (90%)
priorityPercentileBasisPoints = 1/60 (since we want to always get included in the next block)

Here, we would take the last 60 blocks worth of txns. If we're more than 90% full, we pay one fee bucket higher than the "top block"'s txn. Otherwise, we pay one fee bucket lower than that txn. I think that makes sense.
```